### PR TITLE
boot: introduce good recovery systems, provide compatibility handling

### PR DESCRIPTION
**Pull request by bboozzoo**
Introduce good_recover_systems into modeev. Provide basic compatibility handling.

### CODE REVIEW

File: boot/bootstate20.go [hunk 1]
```diff
@@ -754,3 +754,27 @@ func trustedCommandLineBootState(dev Device) *bootState20CommandLine {
...
 	}
 }
+
+// bootState20RecoverySystem implements the successfulBootState interface for
+// tried recovery systems
+type bootState20RecoverySystem struct {
+	dev Device
+}
+
+func (ba20 *bootState20RecoverySystem) markSuccessful(update bootStateUpdate) (bootStateUpdate, error) {
```
Comment (anonymouse64) [inline boot/bootstate20.go:764]:
```suggestion
func (brs20 *bootState20RecoverySystem) markSuccessful(update bootStateUpdate) (bootStateUpdate, error) {
```

?

but again the acronym short hands for all the bootstate things are getting a bit unreadable anyways so maybe it doesn't matter so much

File: boot/modeenv.go [hunk 2]
```diff
@@ -45,16 +45,23 @@ type bootCommandLines []string
 // Modeenv is a file on UC20 that provides additional information
 // about the current mode (run,recover,install)
 type Modeenv struct {
-	Mode                   string   `key:"mode"`
-	RecoverySystem         string   `key:"recovery_system"`
+	Mode           string `key:"mode"`
+	RecoverySystem string `key:"recovery_system"`
+	// CurrentRecoverySystems is a list of labels corresponding to recovery
+	// systems have been verified or are in the process of being tried, thus
```
Comment (anonymouse64) [inline boot/modeenv.go:51]:
```suggestion
	// systems that have been verified or are in the process of being tried, thus
```

File: boot/modeenv_test.go [hunk 3]
```diff
@@ -305,6 +308,24 @@ func (s *modeenvSuite) TestDeepEquals(c *C) {
 	// clear kernel command lines list
 	modeenv2.CurrentKernelCommandLines = nil
 	c.Assert(modeenv1.DeepEqual(modeenv2), Equals, false)
+
```
Comment (anonymouse64) [inline boot/modeenv_test.go:311]: thanks for updating this test

**Review by anonymouse64 — APPROVED** [review]
lgtm, one naming question though that is probably irrelevant

File: boot/boot_test.go [hunk 4]
```diff
@@ -2764,6 +2765,81 @@ func (s *bootenv20Suite) TestMarkBootSuccessful20CommandLineNonRunMode(c *C) {
 	})
 }
 
+func (s *bootenv20Suite) TestMarkBootSuccessful20SystemsCompat(c *C) {
+	b := bootloadertest.Mock("mock", s.bootdir)
+	s.forceBootloader(b)
+
+	// a pending kernel command line change
```
Comment (pedronis) [inline boot/boot_test.go:2772]: is this comment correct?

File: boot/boot_test.go [hunk 5]
```diff
@@ -2764,6 +2765,81 @@ func (s *bootenv20Suite) TestMarkBootSuccessful20CommandLineNonRunMode(c *C) {
...
+	// good recovery systems has been populated
+	c.Check(m2.GoodRecoverySystems, DeepEquals, []string{"1234"})
+	c.Check(m2.CurrentRecoverySystems, DeepEquals, []string{"1234"})
+}
+
+func (s *bootenv20Suite) TestMarkBootSuccessful20SystemsPopulated(c *C) {
+	b := bootloadertest.Mock("mock", s.bootdir)
+	s.forceBootloader(b)
+
+	// a pending kernel command line change
```
Comment (pedronis) [inline boot/boot_test.go:2809]: same

File: boot/boot_test.go [hunk 6]
```diff
@@ -2764,6 +2765,81 @@ func (s *bootenv20Suite) TestMarkBootSuccessful20CommandLineNonRunMode(c *C) {
...
+	coreDev := boottest.MockUC20Device("", nil)
+	c.Assert(coreDev.HasModeenv(), Equals, true)
+	// mark successful
+	err := boot.MarkBootSuccessful(coreDev)
+	c.Assert(err, IsNil)
+
+	// check the modeenv
+	m2, err := boot.ReadModeenv("")
+	c.Assert(err, IsNil)
+	// good recovery systems has been populated
```
Comment (pedronis) [inline boot/boot_test.go:2838]: has not been modified ?

File: boot/modeenv.go (hunk 2 above)
Comment (pedronis) [reply boot/modeenv.go:51]: s/verified/tested/

File: boot/modeenv.go [hunk 7]
```diff
@@ -45,16 +45,23 @@ type bootCommandLines []string
...
 	CurrentRecoverySystems []string `key:"current_recovery_systems"`
-	Base                   string   `key:"base"`
-	TryBase                string   `key:"try_base"`
-	BaseStatus             string   `key:"base_status"`
-	CurrentKernels         []string `key:"current_kernels"`
-	Model                  string   `key:"model"`
-	BrandID                string   `key:"model,secondary"`
-	Grade                  string   `key:"grade"`
+	// GoodRecoverySystems is a list of labels corresponding to recovery
+	// systems were verified and we are prepared to use for recovering. The
```
Comment (pedronis) [inline boot/modeenv.go:55]: s/verified/tested/

File: boot/bootstate20.go (hunk 1 above)
Comment (pedronis) [reply boot/bootstate20.go:764]: I think ba20 comes from bootState20BootAssets.  So yes brs20  here,  and bcl20  for the commandline previously landed would be better.

**Review by pedronis — APPROVED** [review]
thanks,  some small comments,  some bits that  that were reused need some tweaks

**Pull request merged**